Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs

```diff
@@ -211,61 +229,275 @@ public void OnEndHttpCallbackSetsSuccessToFalseForMoreThan400()
-            this.httpProcessingFramework.OnEndHttpCallback(100, null, false, 500);
+            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(this.testUrlNonStandardPort);
+            var id = ClientServerDependencyTracker.GetIdForRequestObject(request);
+            this.httpProcessingFramework.OnRequestSend(request);  
+            this.httpProcessingFramework.OnEndHttpCallback(id, null, false, 500);
 
             Assert.AreEqual(1, this.sendItems.Count, "Exactly one telemetry item should be sent");
             DependencyTelemetry receivedItem = (DependencyTelemetry)this.sendItems[0];
-            string expectedTarget = testUrl.Host + ":" + testUrl.Port;
+            string expectedTarget = this.testUrlNonStandardPort.Host + ":" + this.testUrlNonStandardPort.Port;
             Assert.AreEqual(expectedTarget, receivedItem.Target, "HttpProcessingFramework returned incorrect target for non standard port.");
         }
 
-        #endregion //BeginEndCallBacks
+        /// <summary>
+        /// Validates that even if OnBeginHttpCallback is called after OnRequestSend, the
+        /// packet should match what's written during OnRequestSend.
+        /// </summary>
+        [TestMethod]
+        public void RddTestHttpProcessingFrameworkUpdateTelemetryName()
+        {
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(this.testUrl);
+            var id = ClientServerDependencyTracker.GetIdForRequestObject(request);
+            this.httpProcessingFramework.OnRequestSend(request);  
+            this.httpProcessingFramework.OnBeginHttpCallback(id, this.testUrl.OriginalString);
+            Thread.Sleep(this.sleepTimeMsecBetweenBeginAndEnd);
+            Assert.AreEqual(0, this.sendItems.Count, "No telemetry item should be processed without calling End");
+            this.httpProcessingFramework.OnEndHttpCallback(id, true, false, 200);
+            stopwatch.Stop();
 
-        #region AsyncScenarios
+            Assert.AreEqual(1, this.sendItems.Count, "Only one telemetry item should be sent");
+            ValidateTelemetryPacketForOnRequestSend(this.sendItems[0] as DependencyTelemetry, this.testUrl, RemoteDependencyConstants.HTTP, true, stopwatch.Elapsed.TotalMilliseconds, "200");
+        }
 
         /// <summary>
-        /// Validates HttpProcessingFramework calculates startTime from the start of very first BeginGetRequestStream if any
+        /// Validates that even if multiple events have fired, as long as there is only
+        /// one HttpWebRequest, only one event should be written, during the first call
+        /// to OnEndHttpCallback.
+        /// </summary>
+        [TestMethod]
+        public void RddTestHttpProcessingFrameworkNoDuplication()
+        {
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(this.testUrl);
```

Comment:
Is it possible that customers reuse the same request on purpose, e.g. for some "ping" test or it will be returned from cache and they would need to create a new one each time to go via network? I'm asking to make sure we do not have an edge case here where the behavior should be the opposite (collect all, not the first one).